File: client/round/round.go
package round

import (
	"flare-common/merkle"
	"flare-common/payload"
	"flare-common/policy"
	"fmt"
	"local/fdc/client/attestation"
	bitvotes "local/fdc/client/attestation/bitVotes"
	"local/fdc/client/utils"
	"math/big"
	"sort"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/pkg/errors"
)

type RoundStatus int

const (
	Collecting RoundStatus = iota
	Choosing
	Retry
	Committed
	Done
	Failed
)

type Round struct {
	ID                           uint32
	Attestations                 []*attestation.Attestation
	attestationMap               map[common.Hash]*attestation.Attestation
	bitVotes                     []*bitvotes.WeightedBitVote
	bitVoteCheckList             map[common.Address]*bitvotes.WeightedBitVote
	ConsensusCalculationFinished bool
	ConsensusBitVote             bitvotes.BitVote
	voterSet                     *policy.VoterSet
	merkleTree                   merkle.Tree
}

// New returns a pointer to a new round with ID and voterSet.
func New(ID uint32, voterSet *policy.VoterSet) *Round {
	return &Round{
		ID:                           ID,
		voterSet:                     voterSet,
		attestationMap:               make(map[common.Hash]*attestation.Attestation),
		bitVoteCheckList:             make(map[common.Address]*bitvotes.WeightedBitVote),
		ConsensusCalculationFinished: false,
	}
}

// AddAttestation checks whether an attestation with such request is already in the round.
// If not it is added to the round, if yes the fee is added to the existent attestation
// and Index is set to the earlier one.
func (r *Round) AddAttestation(attToAdd *attestation.Attestation) bool {
	identifier := crypto.Keccak256Hash(attToAdd.Request)
	att, exists := r.attestationMap[identifier]
	if exists {
		att.Fee.Add(att.Fee, attToAdd.Fee)
		if attestation.EarlierLog(attToAdd.Index(), att.Index()) {
			att.Indexes = utils.Prepend(att.Indexes, attToAdd.Index())
		}
		att.Indexes = append(att.Indexes, attToAdd.Index())

		return false
	}

	r.attestationMap[identifier] = attToAdd
	r.Attestations = append(r.Attestations, attToAdd)

	return true
}

// sortAttestations sorts round's attestations according to their IndexLog.
// we assume that attestations have at least one index.
func (r *Round) sortAttestations() {
	sort.Slice(r.Attestations, func(i, j int) bool {
		return attestation.EarlierLog(r.Attestations[i].Index(), r.Attestations[j].Index())
	})
}

// BitVote returns the BitVote for the round according to the current status of Attestations.
func (r *Round) BitVote() (bitvotes.BitVote, error) {
	r.sortAttestations()
	return attestation.BitVoteFromAttestations(r.Attestations)
}

// BitVoteHex returns the 0x prefixed hex string encoded BitVote for the round according to the current status of Attestations.
func (r *Round) BitVoteBytes() ([]byte, error) {
	bitVote, err := r.BitVote()
	if err != nil {
		return nil, fmt.Errorf("cannot get bitVote for round %d: %s", r.ID, err)
	}

	return bitVote.EncodeBitVote(), nil
}

// ComputeConsensusBitVote computes the consensus BitVote according to the collected bitVotes and sets consensus status to the attestations.
func (r *Round) ComputeConsensusBitVote() error {
	defer func() { r.ConsensusCalculationFinished = true }()
	r.sortAttestations()

	fees := make([]*big.Int, len(r.Attestations))
	for i, a := range r.Attestations {
		fees[i] = a.Fee
	}

	consensus, err := bitvotes.EnsembleConsensusBitVote(r.bitVotes, fees, r.voterSet.TotalWeight, 20000000)
	if err != nil {
		return err
	}

	r.ConsensusBitVote = consensus

	return r.setConsensusStatus(consensus)
}

// GetConsensusBitVote returns triplet
// - consensus BitVote
// - bool indicating whether the consensus BitVote is successfully computed
// - bool indicating whether the consensus BitVote computation took place
func (r *Round) GetConsensusBitVote() (bitvotes.BitVote, bool, bool) {
	if r.ConsensusBitVote.BitVector == nil {
		return bitvotes.BitVote{}, false, r.ConsensusCalculationFinished
	}
	return r.ConsensusBitVote, true, r.ConsensusCalculationFinished
}

// ConsensusBitVoteHex returns hex string encoded consensus BitVote if it is already computed.
func (r *Round) ConsensusBitVoteHex() (string, error) {

	if r.ConsensusBitVote.BitVector == nil {
		return "", errors.New("no consensus bitVote")
	}

	return r.ConsensusBitVote.EncodeBitVoteHex(), nil
}

// setConsensusStatus sets consensus status of the attestations.
// The scenario where a chosen attestation is missing is not possible as in such case, it is not possible to compute the consensus bitVote.
// It is assumed that the Attestations are already ordered.
func (r *Round) setConsensusStatus(consensusBitVote bitvotes.BitVote) error {
	// sanity check
	if consensusBitVote.BitVector.BitLen() > len(r.Attestations) {
		return fmt.Errorf("consensus bitVector too long %d", r.ID)
	}

	for i := range r.Attestations {
		r.Attestations[i].Consensus = consensusBitVote.BitVector.Bit(i) == 1
	}

	return nil
}

// MerkleTree computes Merkle tree from sorted hashes of attestations chosen by the consensus bitVote.
// The computed tree is stored in the round.
// If any of the hash of the chosen attestations is not successfully verified, the tree is not computed.
func (r *Round) MerkleTree() (merkle.Tree, error) {
	var hashes []common.Hash
	for i := range r.Attestations {
		if r.Attestations[i].Consensus {
			if r.Attestations[i].Status != attestation.Success {
				return merkle.Tree{}, errors.Errorf("attestation %d in consensus but not confirmed", i)
			}

			hashes = append(hashes, r.Attestations[i].Hash)
		}
	}

	merkleTree := merkle.Build(hashes, false)
	r.merkleTree = merkleTree

	return merkleTree, nil
}

// MerkleTreeCached gets Merkle tree from cache if it is already computed or computes it.
func (r *Round) MerkleTreeCached() (merkle.Tree, error) {
	if len(r.merkleTree) != 0 {

		return r.merkleTree, nil
	}

	return r.MerkleTree()
}

// MerkleRoot returns Merkle root for a round if there is one.
func (r *Round) MerkleRoot() (common.Hash, error) {
	tree, err := r.MerkleTreeCached()
	if err != nil {
		return common.Hash{}, err
	}

	return tree.Root()
}

// ProcessBitVote decodes bitVote message, checks roundCheck, adds voter weight and index, and stores bitVote to the round.
// If the voter is invalid, or has zero weight, the bitVote is ignored.
// If a voter already submitted a valid bitVote for the round, the bitVote is overwritten.
func (r *Round) ProcessBitVote(message payload.Message) error {
	bitVote, err := bitvotes.DecodeBitVoteBytes(message.Payload)
	if err != nil {
		return err
	}

	if int(bitVote.Length) != len(r.Attestations) {
		return fmt.Errorf("got bits %d, have %d attestations", int(bitVote.Length), len(r.Attestations))
	}

	if bitVote.BitVector.BitLen() > len(r.Attestations) {
		return fmt.Errorf("bitVector too long")
	}

	signingAddress, exists := r.voterSet.SubmitToSigningAddress[message.From] //message.From = submit address
	if !exists {
		return fmt.Errorf("no signing address")
	}

	voter, exists := r.voterSet.VoterDataMap[signingAddress]
	if !exists {
		return fmt.Errorf("invalid voter")
	}

	weight := voter.Weight
	if weight <= 0 {
		return fmt.Errorf("zero weight voter")
	}

	// check if a bitVote was already submitted by the sender
	weightedBitVote, exists := r.bitVoteCheckList[message.From]
	if !exists {
		// first submission
		weightedBitVote = &bitvotes.WeightedBitVote{
			BitVote: bitVote,
			Weight:  weight,
			Index:   voter.Index,
			IndexTx: bitvotes.IndexTx{
				BlockNumber:      message.BlockNumber,
				TransactionIndex: message.TransactionIndex,
			},
		}

		r.bitVotes = append(r.bitVotes, weightedBitVote)
	} else if exists && bitvotes.EarlierTx(weightedBitVote.IndexTx, bitvotes.IndexTx{BlockNumber: message.BlockNumber, TransactionIndex: message.TransactionIndex}) {
		// more than one submission. The later submission is considered to be valid.
		weightedBitVote.BitVote = bitVote
		weightedBitVote.Weight = weight
		weightedBitVote.Index = voter.Index
		weightedBitVote.IndexTx = bitvotes.IndexTx{
			BlockNumber:      message.BlockNumber,
			TransactionIndex: message.TransactionIndex,
		}
	}

	return nil
}

File: src/round.py
import hashlib
from typing import List, Dict, Optional, Any

from src.attestation.structures import Attestation, AttestationStatus, RoundStatusMutex, RoundStatus
from src.attestation.bit_votes.structures import BitVote, WeightedBitVote, IndexTx, Value
from src.attestation.bit_votes.coding import decode_bit_vote_bytes
from src.attestation.bit_votes.creation import bit_vote_from_attestations
from src.attestation.bit_votes.consensus import branch_and_bound_bits_double
from src.attestation.bit_votes.preprocessing import filter_and_aggregate, assemble_solution
from src.attestation.verification import Request

# --- Mock Objects for External Dependencies ---

class MockVoter:
    """A mock to stand in for the complex VoterData object."""
    def __init__(self, index: int, weight: int):
        self.index = index
        self.weight = weight

class MockVoterSet:
    """A mock to stand in for the complex VoterSet object."""
    def __init__(self, voters: Dict[str, MockVoter], total_weight: int):
        # Maps signing address to voter data
        self.voter_data_map = voters
        self.total_weight = total_weight
        # Maps submit address to signing address
        self.submit_to_signing_address = {f"submit_{k}": k for k in voters.keys()}

# --- Round Class ---

class Round:
    """
    Represents all data and logic for a single voting round.
    This is a Python port of the `Round` struct and its methods from `round.go`.
    """
    def __init__(self, round_id: int, voter_set: MockVoterSet):
        self.id = round_id
        self.status = RoundStatusMutex(value=RoundStatus.PRE_CONSENSUS)
        self.voter_set = voter_set

        self.attestations: List[Attestation] = []
        self.attestation_map: Dict[bytes, Attestation] = {}

        self.bit_votes: List[WeightedBitVote] = []
        self.bit_vote_checklist: Dict[str, WeightedBitVote] = {} # Keyed by submit address

        self.consensus_calculation_finished = False
        self.consensus_bit_vote: Optional[BitVote] = None
        self.merkle_tree: Optional[Any] = None # Placeholder for a Merkle tree object

    def add_attestation(self, attestation: Attestation) -> bool:
        """
        Adds an attestation to the round, handling duplicates.
        Returns True if the attestation was new, False if it was a duplicate.
        """
        # Using a hash of the request as a unique identifier
        identifier = hashlib.sha256(attestation.request).digest()

        if identifier in self.attestation_map:
            existing_att = self.attestation_map[identifier]
            existing_att.fee += attestation.fee
            # This logic for prepending/appending indexes can be simplified
            # for the port, as long as sorting happens before use.
            existing_att.indexes.extend(attestation.indexes)
            return False

        self.attestation_map[identifier] = attestation
        self.attestations.append(attestation)
        attestation.round_status = self.status
        return True

    def _sort_attestations(self):
        """Sorts attestations by their index log (block number, then log index)."""
        self.attestations.sort(key=lambda a: (a.get_first_index().block_number, a.get_first_index().log_index))

    def get_local_bit_vote(self) -> BitVote:
        """Generates the bit vote from this node's perspective."""
        self._sort_attestations()
        return bit_vote_from_attestations(self.attestations)

    def process_incoming_bit_vote(self, message: Dict) -> None:
        """
        Processes a bit vote message from another participant.
        A port of `ProcessBitVote`.
        """
        # In Go, message is a complex struct. Here we use a simple dict.
        submit_address = message["from"]

        # Look up voter info
        signing_address = self.voter_set.submit_to_signing_address.get(submit_address)
        if not signing_address:
            raise ValueError(f"No signing address for submit address {submit_address}")

        voter = self.voter_set.voter_data_map.get(signing_address)
        if not voter or voter.weight <= 0:
            raise ValueError(f"Invalid or zero-weight voter for address {signing_address}")

        # Decode the bit vote from the payload
        bit_vote = decode_bit_vote_bytes(message["payload"])
        if bit_vote.length != len(self.attestations):
            raise ValueError("Bit vote length mismatch")

        # Create the weighted bit vote
        new_wbv = WeightedBitVote(
            index=voter.index,
            weight=voter.weight,
            bit_vote=bit_vote,
            index_tx=IndexTx(message["block_number"], message["tx_index"])
        )

        # Handle re-submissions
        if submit_address not in self.bit_vote_checklist:
            self.bit_votes.append(new_wbv)
            self.bit_vote_checklist[submit_address] = new_wbv
        else:
            # Overwrite if the new one is later
            existing_wbv = self.bit_vote_checklist[submit_address]
            if not (existing_wbv.index_tx.block_number > new_wbv.index_tx.block_number or \
               (existing_wbv.index_tx.block_number == new_wbv.index_tx.block_number and \
                existing_wbv.index_tx.transaction_index > new_wbv.index_tx.transaction_index)):

                # Find and replace in the main list
                for i, bv in enumerate(self.bit_votes):
                    if bv.index == existing_wbv.index:
                        self.bit_votes[i] = new_wbv
                        break
                self.bit_vote_checklist[submit_address] = new_wbv

    def compute_consensus(self, max_ops: int = 20_000_000):
        """
        Runs the full consensus algorithm on the collected bit votes.
        A port of `ComputeConsensusBitVote`.
        """
        self._sort_attestations()

        fees = [a.fee for a in self.attestations]

        # 1. Preprocessing
        agg_votes, agg_bits, fr = filter_and_aggregate(
            self.bit_votes, fees, self.voter_set.total_weight
        )

        # 2. Branch and Bound
        solution = branch_and_bound_bits_double(
            agg_votes, agg_bits, fr.guaranteed_weight,
            self.voter_set.total_weight, fr.guaranteed_fees, max_ops, Value(0,0)
        )

        # 3. Assemble the final bit vote
        self.consensus_bit_vote = assemble_solution(fr, solution.bits, len(self.attestations))
        self.consensus_calculation_finished = True

        # 4. Update the status of all attestations
        self.status.value = RoundStatus.CONSENSUS
        self._set_consensus_status()

    def _set_consensus_status(self):
        """Updates the 'consensus' flag on each attestation based on the result."""
        if self.consensus_bit_vote is None:
            return

        for i, att in enumerate(self.attestations):
            if (self.consensus_bit_vote.bit_vector >> i) & 1:
                att.consensus = True
            else:
                att.consensus = False

    def get_merkle_root(self) -> Optional[bytes]:
        """
        Builds a Merkle tree from confirmed attestations and returns the root.
        This is a simplified version of the Go implementation.
        """
        if self.status.value != RoundStatus.CONSENSUS:
            raise RuntimeError("Cannot build Merkle tree before consensus is reached.")

        hashes = []
        for att in self.attestations:
            if att.consensus:
                if att.status != AttestationStatus.SUCCESS or att.hash is None:
                    # In a real system, this is a critical error.
                    raise ValueError(f"Attestation in consensus but not confirmed: {att}")
                hashes.append(att.hash)

        if not hashes:
            self.merkle_tree = b'\x00' * 32 # Placeholder
            return self.merkle_tree

        # A real implementation would use a proper Merkle library.
        # For this test, we'll just hash the concatenation of hashes.
        import hashlib
        m = hashlib.sha256()
        for h in sorted(hashes): # Sort for determinism
            m.update(h)

        self.merkle_tree = m.digest()
        self.status.value = RoundStatus.DONE
        return self.merkle_tree